README.md
# [witches] Witches (Mobs Redo addon) [mod] for Minetest 

Witches is copyright 2022 Francisco Athens, Ramona Athens, Damon Athens and Simone Athens
The MIT License (MIT)

* code/issue tracking :https://gitlab.com/freelikegnu/witches
* forum discussion: https://forum.minetest.net/viewtopic.php?t=25084
* ContentDB: https://content.minetest.net/packages/FreeLikeGNU/witches/
  
Witches inhabit the land! They are currently in development but, can already:

* Spawn near wood above ground and in their cottages.
  
* Witches have magic!  They will defend themselves and each other against aggressive players and mobs!
  
* If you are repeatedly helpful to a witch something good is likely to happen!
  
* New Witch cottage code with no additional mod requirements!

* If protector (or similar) mod is available and enabled, Witches will not build in these areas! 
  
* Chance of cottage to spawn over a dungeon if they are near the surface.

* Have a name and origin location which they will tell player.

* Randomly choose hair color and style

* Randomly choose hat style and hat ornaments

* Basic interactivity:
    will follow player for holding certain items.
    one of those items they will request with randomly chosen text, if player right clicks
    if player is holding that requested item and right click witch player will receive a reward!
    the reward is a personalized tool with increased stats!

## Required Mods:
* Mobs Redo by TenPlus1 API as of version 20200516: to run
    * https://forum.minetest.net/viewtopic.php?f=9&t=9917
    * Mobs Redo git repository:  https://notabug.org/TenPlus1/mobs_redo

## Optional Mods:
* protector (or similar) - to prevent Witches from placing cottages in protected areas! 
  
* Any mobs from any mobs mod attacking Witches may be turned into a sheep!

* Fireflies (Minetest Game) for certain effects and rewards
  
## Licenses of Source Media Files:
* witches_witch.b3d and witch.blend 
    * copyright 2020 Francisco Athens, Ramona Athens, Damon Athens and Simone Athens
      Creative Commons Attribution-ShareAlike 3.0 Unported (CC BY-SA 3.0)
    * http://creativecommons.org/licenses/by-sa/3.0/

* above meshes based on character from minetest_game
    * by MirceaKitsune (WTFPL)
    * https://github.com/minetest/minetest_game/blob/master/mods/default/README.txt#L71

* witches_*.png files and witches_*.xcf files:
    * copyright 2020 Francisco Athens, Ramona Athens, Damon Athens and Simone Athens
      Creative Commons  Attribution-ShareAlike 3.0 Unported (CC BY-SA 3.0) 
    * http://creativecommons.org/licenses/by-sa/3.0/

* sheep.blend, sheep.xcf witches_sheep.b3d, witches_sheep.png witches_sheep_wool.png:
    * copyright 2020 Francisco Athens
      Creative Commons  Attribution-ShareAlike 3.0 Unported (CC BY-SA 3.0) 
    * http://creativecommons.org/licenses/by-sa/3.0/

* sheep.lua code adapted from mobs_animals by TenPlus1 and authors credited in sheep.lua
  
* Sound files:
    witches_magic01.ogg  MATRIXXX_ December 28th, 2018 CC-BY 3.0
      https://freesound.org/people/MATRIXXX_/sounds/455205/

    witches_magic02.ogg RICHERlandTV January 28th, 2014 CC-BY 3.0
      https://freesound.org/people/RICHERlandTV/sounds/216089/

    witches_sheep.ogg - zachrau March 6th, 2017 CC0
      https://freesound.org/people/zachrau/sounds/383144/
    
    witches_water.ogg - InspectorJ  March 10th, 2018 CC BY 3.0 
      https://freesound.org/people/InspectorJ/sounds/421184/
   
  Thanks to TenPlus1 for keeping the Mobs_Redo going!
  Thanks to rubenwardy for awesome help and the Minetest ContentDB
  Thanks to everyone in the Minetest forums and IRC for just being great!
  Thanks to ctate for naming the witches first spell "Yeet"!

GET MINETEST: https://www.minetest.net/


changelog
    2022/10/16:  Cottage chest container items and extra furniture moved to database!
                  Note: Cottages already generated will not be affected.
    2022/10/11: Fix init and quests again!
    2022/10/09B: Fixed special gift frequency and added debug categories
    2022/10/09: Issue 5: invalid 'random' (interval is empty): check before crash.
    2022/10/08: moved quest item lists out of code and into database for server customizations
    2022/10/05: quest ui checks
    2022/10/04: checks for valid math.random(), less frequent cottage tree and bush growth (abms), ui adjustments 
    2022/10/02: Quests can now require higher counts of items to complete, more cottage fixing
    2022/09/30: add a debug check for annoying bug
    2022/09/04: More cottage code cleanup and re-add witch spawner (sometimes it works!?)
    2022/09/01: Fix debug crashing on older MT (at least from 5.4)
    2022/08/28: More Cottage variations and furniture, cleanup of node workarounds, fixes to witch spawning in cottages 
    2022/08/27: Clean up deprecated vector tables, quest and chest fixes
    2022/02/23: Fixes and formatting
    2022/02/22: Restored Cottages spawning over dungeons (!) and some related settings.
    2022/02/19: Require only default and mobs now, additional checks, cottage generation now optional from game Settings tab > All Settings > Mods > witches
    2022/02/14: Dialog updates
    2022/02/13: Debug output can be enabled from game Settings tab > All Settings > Mods > witches 
    2022/02/12: Witches honor protected nodes ("protector" mod support)
    2020/08/10: Testing new cottage building code.
    2020/08/06: Fixed witches not registered when handle_schematics is not loaded! D'oh!
    2020/08/01: Witches have two new wands from Damon and Simone!
    2020/07/31: Witches can turn aggressive mobs into sheep! Witches gain drench spell!
    2020/07/28: Witches will teleport you away from themselves if you are aggressive!
    2020/07/26: Witch cottage placement refactor and settings menu options!
    2020/07/25: Witch cottages refinements (thanks Solomine!)
    2020/07/20: Witches have homes (modified procedural Basic_Houses script by Sokomine)! (If Sokomine's handle_schematics mod is also active) L-System trees are added near witch cottages too!!
      video: https://youtu.be/qDS3fT-ROR0
    2020/07/19: Hairstyles are now randomly chosen, animated hat test.
    Update 2020/07/17: More Quest UI work (reward items try to fit into player inv or drop) and some variations in quest text and hats.
    video: https://youtu.be/Lor06w5IMIQ


sheep.lua
--Code based on mobs_animal "sheep" 
--Copyright (c) 2014 Krupnov Pavel and 2016 TenPlus1 The MIT License (MIT)

local variance = witches.variance
local S = mobs.intllib

witches.sheep_colors = {
	{"black",      S("Black"),      "#000000b0"},
	{"blue",       S("Blue"),       "#015dbb70"},
	{"brown",      S("Brown"),      "#663300a0"},
	{"cyan",       S("Cyan"),       "#01ffd870"},
	{"dark_green", S("Dark Green"), "#005b0770"},
	{"dark_grey",  S("Dark Grey"),  "#303030b0"},
	{"green",      S("Green"),      "#61ff0170"},
	{"grey",       S("Grey"),       "#5b5b5bb0"},
	{"magenta",    S("Magenta"),    "#ff05bb70"},
	{"orange",     S("Orange"),     "#ff840170"},
	{"pink",       S("Pink"),       "#ff65b570"},
	{"red",        S("Red"),        "#ff0000a0"},
	{"violet",     S("Violet"),     "#2000c970"},
	{"white",      S("White"),      "#abababc0"},
	{"yellow",     S("Yellow"),     "#e3ff0070"},
}

local all_colours = witches.sheep_colors

-- Sheep model and textures by FreeLikeGNU 2020 CC-BY-SA-3.0

for _, col in ipairs(all_colours) do

	mobs:register_mob("witches:sheep_"..col[1], {
		stay_near = {"farming:straw", 10},
		stepheight = 0.6,
		type = "animal",
		passive = true,
		hp_min = 8,
		hp_max = 10,
		armor = 200,
		collisionbox = {-0.5, 0, -0.5, 0.5, 1, 0.5},
		visual = "mesh",
		mesh = "witches_sheep.b3d",
		textures = {
			{"witches_sheep.png^(witches_sheep_wool.png^[colorize:" .. col[3]..")"},
		},
		--gotten_texture = {"mobs_sheep_shaved.png"},
		--gotten_mesh = "mobs_sheep_shaved.b3d",
		makes_footstep_sound = true,
		sounds = {
			random = "witches_sheep",
		},
		walk_velocity = 1,
		run_velocity = 2,
		runaway = true,
		jump = true,
		jump_height = 6,
		pushable = true,
		drops = {
			{name = "mobs:mutton_raw", chance = 1, min = 1, max = 2},
			{name = "wool:"..col[1], chance = 1, min = 1, max = 1},
		},
		water_damage = 0,
		lava_damage = 5,
		light_damage = 0,
		animation = {
			speed_normal = 40,
			speed_run = 80,
			stand_start = 1,
			stand_end = 60,
			walk_start = 70,
			walk_end = 90,
		},
		follow = {"farming:wheat", "default:grass_1"},
		view_range = 8,
		replace_rate = 10,
		replace_what = {
			{"group:grass", "air", -1},
			{"default:dirt_with_grass", "default:dirt", -2}
		},
		fear_height = 3,

		on_spawn = function(self)
			if not self.size then
				local overall = variance(100,200)
				self.size = vector.new(overall*variance(90,110), overall*variance(90,110), overall*variance(90,110))
			end

			local self_properties = self.object:get_properties()
			self_properties.visual_size = self.size
			self.object:set_properties(self_properties)
		end,

		on_replace = function(self, pos, oldnode, newnode)

			self.food = (self.food or 0) + 1

			-- if sheep replaces 8x grass then it regrows wool
			if self.food >= 8 then

				self.food = 0
				self.gotten = false

				self.object:set_properties({
					textures = {"witches_sheep.png^(witches_sheep_wool.png^[colorize:" .. col[3]..")"},
					mesh = "witches_sheep.b3d",
				})
			end
		end,
		on_rightclick = function(self, clicker)

			--are we feeding?
			if mobs:feed_tame(self, clicker, 8, true, true) then

				--if fed 7x grass or wheat then sheep regrows wool
				if self.food and self.food > 6 then

					self.gotten = false

					self.object:set_properties({
						textures = {"witches_sheep.png^(witches_sheep_wool.png^[colorize:" .. col[3]..")"},
						mesh = "witches_sheep.b3d",
					})
				end

				return
			end

			local item = clicker:get_wielded_item()
			local itemname = item:get_name()
			local name = clicker:get_player_name()

			--are we giving a haircut>
			if itemname == "mobs:shears" then

				if self.gotten ~= false
				or self.child ~= false
				or name ~= self.owner
				or not minetest.get_modpath("wool") then
					return
				end

				self.gotten = true -- shaved

				local obj = minetest.add_item(
					self.object:get_pos(),
					ItemStack( "wool:" .. col[1] .. " " .. math.random(1, 3) )
				)

				if obj then

					obj:setvelocity({
						x = math.random(-1, 1),
						y = 5,
						z = math.random(-1, 1)
					})
				end

				item:add_wear(650) -- 100 uses

				clicker:set_wielded_item(item)

				self.object:set_properties({
					textures = {"witches_sheep.png"}, --should be shaved
					mesh = "witches_sheep.b3d",
				})

				return
			end

			--are we coloring?
			if itemname:find("dye:") then

				if self.gotten == false
				and self.child == false
				and self.tamed == true
				and name == self.owner then

					local colr = string.split(itemname, ":")[2]

					for _,c in pairs(all_colours) do

						if c[1] == colr then

							local pos = self.object:get_pos()

							self.object:remove()

							local mob = minetest.add_entity(pos, "witches:sheep_" .. colr)
							local ent = mob:get_luaentity()

							ent.owner = name
							ent.tamed = true

							-- take item
							if not mobs.is_creative(clicker:get_player_name()) then
								item:take_item()
								clicker:set_wielded_item(item)
							end

							break
						end
					end
				end

				return
			end

			-- protect mod with mobs:protector item
			if mobs:protect(self, clicker) then return end

			--are we capturing?
			if mobs:capture_mob(self, clicker, 0, 5, 60, false, nil) then return end
		end
	})

mobs:register_egg("witches:sheep_"..col[1], S("@1 Sheep", col[2]), "wool_"..col[1]..".png")

end


mobs:spawn({
	name = "witches:sheep_white",
	nodes = {"default:dirt_with_grass", "ethereal:green_dirt"},
	neighbors = {"group:grass"},
	min_light = 14,
	interval = 60,
	chance = 8000, -- 15000
	min_height = 0,
	max_height = 200,
	day_toggle = true,
})


--mobs:alias_mob("witches:sheep", "mobs_animal:sheep_white") -- compatibility

-- raw mutton
minetest.register_craftitem(":mobs:mutton_raw", {
	description = S("Raw Mutton"),
	inventory_image = "mobs_meat_raw.png",
	on_use = minetest.item_eat(2),
	groups = {food_meat_raw = 1, food_mutton_raw = 1, flammable = 2},
})

-- cooked mutton
minetest.register_craftitem(":mobs:mutton_cooked", {
	description = S("Cooked Mutton"),
	inventory_image = "mobs_meat.png",
	on_use = minetest.item_eat(6),
	groups = {food_meat = 1, food_mutton = 1, flammable = 2},
})

minetest.register_craft({
	type = "cooking",
	output = "mobs:mutton_cooked",
	recipe = "mobs:mutton_raw",
	cooktime = 5,
})
